blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java

```diff
@@ -0,0 +1,343 @@
+import io.jenkins.blueocean.rest.model.BluePipelineStep;
+import org.jenkinsci.plugins.workflow.actions.NotExecutedNodeAction;
+import org.jenkinsci.plugins.workflow.cps.nodes.StepAtomNode;
+import org.jenkinsci.plugins.workflow.cps.nodes.StepEndNode;
+import org.jenkinsci.plugins.workflow.graph.BlockEndNode;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.graphanalysis.DepthFirstScanner;
+import org.jenkinsci.plugins.workflow.graphanalysis.ForkScanner;
+import org.jenkinsci.plugins.workflow.graphanalysis.MemoryFlowChunk;
+import org.jenkinsci.plugins.workflow.graphanalysis.StandardChunkVisitor;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.jenkinsci.plugins.workflow.pipelinegraphanalysis.GenericStatus;
+import org.jenkinsci.plugins.workflow.pipelinegraphanalysis.StageChunkFinder;
+import org.jenkinsci.plugins.workflow.pipelinegraphanalysis.StatusAndTiming;
+import org.jenkinsci.plugins.workflow.pipelinegraphanalysis.TimingInfo;
+import org.jenkinsci.plugins.workflow.support.actions.PauseAction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * @author Vivek Pandey
+ */
+public class PipelineNodeGraphVisitor extends StandardChunkVisitor implements NodeGraphBuilder{
+    private final WorkflowRun run;
+
+    public final Map<FlowNodeWrapper, List<FlowNodeWrapper>> parentToChildrenMap = new LinkedHashMap<>();
+
+    private final ArrayDeque<FlowNodeWrapper> parallelBranches = new ArrayDeque<>();
+
+    public final ArrayDeque<FlowNodeWrapper> nodes = new ArrayDeque<>();
+
+    private FlowNode firstExecuted = null;
+
+    private FlowNodeWrapper nextStage;
+
+    private FlowNode branchEnd;
+
+    private FlowNode parallelEnd;
+
+    public final Map<String, FlowNodeWrapper> nodeMap = new HashMap<>();
+
+    private static final Logger logger = LoggerFactory.getLogger(PipelineNodeGraphVisitor.class);
+
+    public PipelineNodeGraphVisitor(WorkflowRun run) {
+        this.run = run;
+        if(run.getExecution()!=null) {
+            ForkScanner.visitSimpleChunks(run.getExecution().getCurrentHeads(), this, new StageChunkFinder());
+        }
+    }
+
+    @Override
+    public void chunkStart(@Nonnull FlowNode startNode, @CheckForNull FlowNode beforeBlock, @Nonnull ForkScanner scanner) {
+        super.chunkStart(startNode, beforeBlock, scanner);
+        if(isNested()){
+            return;
+        }
+
+        if (NotExecutedNodeAction.isExecuted(startNode)) {
+            firstExecuted = startNode;
+        }
+        dump(String.format("chunkStart=> id: %s, name: %s, function: %s", startNode.getId(),
```

Comment:
🐛  The calls to getDisplayName & functionName can be somewhat expensive in bulk.  If you want dumping on the main code path, it should be behind a conditional check if we are doing debug logging before fetching this.